src/thunderbots/software/ai/hl/stp/tactic/tactic.cpp

```diff
@@ -29,57 +26,50 @@ void Tactic::updateRobot(const Robot &robot)
 
 std::unique_ptr<Intent> Tactic::getNextIntent()
 {
+    std::unique_ptr<Intent> next_intent = nullptr;
     if (!robot)
     {
         LOG(WARNING) << "Requesting the next Intent for a Tactic without a Robot assigned"
                      << std::endl;
-        return std::unique_ptr<Intent>{};
     }
-
-    auto next_intent = getNextIntentHelper();
-    if (done_ && loop_forever)
+    else
     {
-        // If the tactic is done and is supposed to loop forever, we re-create the
-        // intent_sequence which "restarts" the coroutine. We then run the coroutine
-        // again, and return the result from the restarted coroutine rather than the
-        // old one. This way, any callers of this function won't accidentally get a
-        // nullptr returned for a single call (which could come from the "old" coroutine)
-        // when this Tactic restarts
-        intent_sequence.reset();
-        intent_sequence = std::make_unique<intent_coroutine::pull_type>(
-            [this](intent_coroutine::push_type &yield) {
-                return this->calculateNextIntentWrapper(yield);
-            });
         next_intent = getNextIntentHelper();
+        if (done_ && loop_forever)
+        {
+            // Re-start the intent sequence by re-creating it
+            intent_sequence = IntentCoroutine::pull_type(
+                boost::bind(&Tactic::calculateNextIntentWrapper, this, _1));
+            next_intent = getNextIntentHelper();
+        }
     }
 
     return next_intent;
 }
 
-std::unique_ptr<Intent> Tactic::calculateNextIntentWrapper(
-    intent_coroutine::push_type &yield)
+void Tactic::calculateNextIntentWrapper(IntentCoroutine::push_type &yield)
 {
     // Yield a null pointer the very first time the function is called. This value will
     // never be seen/used by the rest of the system.
     yield(std::unique_ptr<Intent>{});
 
     // Anytime after the first function call, the calculateNextIntent function will be
-    // used to perform the real logic
-    return calculateNextIntent(yield);
+    // used to perform the real logic. The calculateNextIntent function will yield its
+    // values to the top of the coroutine stack, where they will be retrieved by
+    // getNextIntent, so we do not need to yield or return the result of this function
+    calculateNextIntent(yield);
 }
 
 std::unique_ptr<Intent> Tactic::getNextIntentHelper()
 {
-    std::unique_ptr<Intent> next_intent;
+    std::unique_ptr<Intent> next_intent = nullptr;
     // Check if the coroutine "iterator" has any more work to do. Only run the coroutine
     // if there is work to be done otherwise the coroutine library will fail on an assert.
```

Comment:
We should add a note here that this runs the coroutine and then checks it.

Reply:
fixed in all locations